src/main/scala/com/fulcrumgenomics/umi/CorrectUmis.scala

```diff
@@ -0,0 +1,229 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.fulcrumgenomics.umi
+
+import com.fulcrumgenomics.FgBioDef._
+import com.fulcrumgenomics.cmdline.{ClpGroups, FgBioTool}
+import com.fulcrumgenomics.util._
+import dagr.commons.util.LazyLogging
+import dagr.sopt.{arg, clp}
+import htsjdk.samtools.{SAMFileWriterFactory, SamReaderFactory}
+
+import scala.collection.mutable
+import CorrectUmis._
+
+object CorrectUmis {
+  /**
+    * Holds information about the match of a UMI to an expected UMI.
+    * @param matched true if the match is acceptable, false otherwise
+    * @param umi the fixed UMI sequence that was the closest match
+    * @param mismatches the number of mismatches between the UMI and the reported best matching fixed UMI
+    *  */
+  private[umi] case class UmiMatch(matched: Boolean, umi: String, mismatches: Int)
+
+  /**
+    * Metrics for the correction of UMI sequences to a fixed set of known UMIs.
+    *
+    * @param umi the corrected UMI sequence (or all Ns for unmatched)
+    * @param total_matches the number of UMI sequences that matched/were corrected to this UMI
+    * @param perfect_matches the number of UMI sequences that were perfect matches to this UMI
+    * @param one_mismatch_matches the number of UMI sequences that matched with a single mismatch
+    * @param two_mismatch_matches the number of UMI sequences that matched with two mismatches
+    * @param other_matches the number of UMI sequences that matched with > 2 mismatches
+    * @param fraction_of_matches the fraction of all UMIs that matched or were corrected to this UMI
+    * @param representation the [[total_matches]] for this UMI divided by the mean [[total_matches]] for all UMIs
+    */
+  case class UmiCorrectionMetrics(umi: String,
+                                  var total_matches: Long = 0,
+                                  var perfect_matches: Long = 0,
+                                  var one_mismatch_matches: Long = 0,
+                                  var two_mismatch_matches: Long = 0,
+                                  var other_matches: Long = 0,
+                                  var fraction_of_matches: Double = 0,
+                                  var representation: Double = 0
+                                 ) extends Metric
```

Comment:
In the case of two or more UMIs per template, I'd like to see a metric per UMI as well as an overall metric for the template/read